pkg/pool-hooks/contracts/StableSurgeHook.sol

```diff
@@ -0,0 +1,244 @@
+import { VaultGuard } from "@balancer-labs/v3-vault/contracts/VaultGuard.sol";
+import { BaseHooks } from "@balancer-labs/v3-vault/contracts/BaseHooks.sol";
+import {
+    LiquidityManagement,
+    TokenConfig,
+    PoolSwapParams,
+    HookFlags,
+    SwapKind
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { StableMath } from "@balancer-labs/v3-solidity-utils/contracts/math/StableMath.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+import { Authentication } from "@balancer-labs/v3-solidity-utils/contracts/helpers/Authentication.sol";
+
+import { StableSurgeMedianMath } from "./utils/StableSurgeMedianMath.sol";
+
+/**
+ * @notice Hook that charges a fee on trades that push a pool into an imbalanced state beyond a given threshold.
+ * @dev Uses the dynamic fee mechanism to apply a directional fee.
+ */
+contract StableSurgeHook is BaseHooks, VaultGuard, Authentication {
+    using FixedPoint for uint256;
+
+    /**
+     * @notice A new `StableSurgeHookExample` contract has been registered successfully.
+     * @dev If the registration fails the call will revert, so there will be no event.
+     * @param pool The pool on which the hook was registered
+     * @param factory The factory that registered the pool
+     */
+    event StableSurgeHookExampleRegistered(address indexed pool, address indexed factory);
+
+    /**
+     * @notice The threshold percentage has been changed for a pool in a `StableSurgeHookExample` contract.
+     * @dev Note, the initial threshold percentage is set on deployment and an event is emitted.
+     * @param pool The pool for which the threshold percentage has been changed
+     * @param newSurgeThresholdPercentage The new threshold percentage
+     */
+    event ThresholdSurgePercentageChanged(address indexed pool, uint256 newSurgeThresholdPercentage);
+
+    /**
+     * @notice A new authorizer is set by `setAuthorizer`.
+     * @param newAuthorizer The address of the new authorizer
+     */
+    event AuthorizerChanged(IAuthorizer indexed newAuthorizer);
+
+    /// @notice The threshold must be a valid percentage value.
+    error InvalidSurgeThresholdPercentage();
+
+    uint256 public constant MAX_SURGE_FEE_PERCENTAGE = 95e16; // 95%
+
+    // The default threshold, above which surging will occur.
+    uint256 private immutable _defaultSurgeThresholdPercentage;
+
+    // Only pools from a specific factory (e.g., StablePoolFactory) are able to register and use this hook.
+    address private immutable _allowedFactory;
+
+    // Upgradeable contract in charge of setting permissions.
+    IAuthorizer internal _authorizer;
+
+    // Store the current threshold for each pool.
+    mapping(address pool => uint256 threshold) private _surgeThresholdPercentage;
+
+    constructor(
+        IVault vault,
+        address allowedFactory,
+        IAuthorizer newAuthorizer,
```

Comment:
Why would we want to set an independent authorizer? Can just use the Vault one (see #1125). I suppose if you wanted to completely generalize it to use a different kind of governance - but the interface is pretty specific, and this is a Balancer hook.